src/control/server/storage/scm/ipmctl.go

```diff
@@ -112,12 +113,68 @@ func (r *cmdRunner) Discover() (storage.ScmModules, error) {
 			SocketID:        uint32(d.Socket_id),
 			PhysicalID:      uint32(d.Physical_id),
 			Capacity:        d.Capacity,
+			UID:             d.Uid.String(),
 		})
 	}
 
 	return modules, nil
 }
 
+func scmFirmwareUpdateStatusFromIpmctl(ipmctlStatus uint32) storage.ScmFirmwareUpdateStatus {
+	switch ipmctlStatus {
+	case ipmctl.FWUpdateStatusFailed:
+		return storage.ScmUpdateStatusFailed
+	case ipmctl.FWUpdateStatusSuccess:
+		return storage.ScmUpdateStatusSuccess
+	case ipmctl.FWUpdateStatusStaged:
+		return storage.ScmUpdateStatusStaged
+	}
+	return storage.ScmUpdateStatusUnknown
+}
+
+func uidStringToIpmctl(uidStr string) (ipmctl.DeviceUID, error) {
+	var uid ipmctl.DeviceUID
+	n := copy(uid[:], uidStr)
+	if n == 0 {
+		return ipmctl.DeviceUID{}, errors.New("invalid SCM module UID")
+	}
+	return uid, nil
+}
+
+// GetFirmwareStatus gets the current firmware status for a specific device.
+func (r *cmdRunner) GetFirmwareStatus(deviceUID string) (*storage.ScmFirmwareInfo, error) {
+	uid, err := uidStringToIpmctl(deviceUID)
+	if err != nil {
+		return nil, errors.New("invalid SCM module UID")
+	}
+	info, err := r.binding.GetFirmwareInfo(uid)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get firmware info for device %q", deviceUID)
+	}
+
+	return &storage.ScmFirmwareInfo{
+		ActiveVersion:     info.ActiveFWVersion.String(),
+		StagedVersion:     info.StagedFWVersion.String(),
+		ImageMaxSizeBytes: info.FWImageMaxSize,
+		UpdateStatus:      scmFirmwareUpdateStatusFromIpmctl(info.FWUpdateStatus),
+	}, nil
+}
+
+// UpdateFirmware attempts to update the firmware on the given device with the binary at
+// the path provided.
+func (r *cmdRunner) UpdateFirmware(deviceUID string, firmwarePath string) error {
+	uid, err := uidStringToIpmctl(deviceUID)
+	if err != nil {
+		return errors.New("invalid SCM module UID")
+	}
+	// Force option permits minor version downgrade.
+	err = r.binding.UpdateFirmware(uid, firmwarePath, true)
```

Comment:
are we sure we want to allow downgrades by default?

Reply:
This was an open question on my end, too. SCM downgrade support is pretty limited, but we can in theory roll back a minor version upgrade. I guess the question is we whether we want to expose the force option further up the stack.